### ISSUE TITLE
IP Groups: Define and enforce IP Group scope of permissions in application

### ISSUE THREAD

**Issue by kcondon** [IQSS/dataverse#1380]
The low level IP Groups infrastructure has been created but the application currently does not check the perms nor enable actions in the application based on them. The groups do appear in the user list in permissions but assigning to roles there or via API has no effect.

Also, there has been some discussion whether IP group permissions should be limited to viewing/downloading rather than writing.

**Comment by michbarsinai** [comment]
Consider merge of permissionsFor and permissionsForUser in the permissionServiceBean

**Comment by kcondon** [comment]
This is also true for explicit groups.

**Comment by michbarsinai** [comment]
Validated by letting localhosts edit a dataset, viewing the dataset while not logging in, then removing the role assignment and trying to view the dataset again. On the second round, got redirected to the login page.
I didn't see the dataset metadata edit controls, but that might be another issue.

**Comment by scolapasta** [comment]
@michbarsinai definitely we should merge them - is that a straightforward thing to do now?

**Comment by kcondon** [comment]
In speaking with Gustavo, there is a requirement that is unique to IP Groups: they cannot edit anything, change anything, effectively not write anything. Rather, this group's main purpose is view and access restricted files and unpublished dv's and ds. So, I will test permissions for this group accordingly.

**Comment by kcondon** [comment]
Issues I have seen with IP Groups that may or may not need fixing, depending on a design/functionality review:
1. Can't assign a single ip address without specifying it as a range of 1: [1.1.1.1-1.1.1.1] (not really a major issue, just making note of it here).
2. Can assign roles that have write permissions to IP groups, eg. admin, editor, though ipGroups are not supposed to perform these functions and in some cases, ie. edit buttons, can't see them due to check on is authenticated user.
3. Can't see unpublished dvs and ds in search results, though can browse to them.
4. Have not tried accessing restricted files since that currently is not working.

I am restricting the scope of this ticket back to being ipGroups only.

**Comment by kcondon** [comment]
This was requested by Open Scholar and is a deal breaker for upgrading to 4.x. See RT 225414:
Oh also, I should have mentioned that support for IP groups is a real
deal-breaker for OCUL. We won't be able to even consider upgrading our
instance without that functionality!

**Comment by michbarsinai** [comment]
This should be supported now, after #1257 is implemented.

**Comment by kcondon** [comment]
Can create ip groups and assign roles but membership does not seem to be detected, eg, using ipGroup3.json, assign member role to an unpublished dv, ds, cannot browse, view, directly access or access via api get dataset.

**Comment by kcondon** [comment]
@michbarsinai when you get a chance, can you take a look? Seems like a regression.

**Comment by michbarsinai** [comment]
The UI has an issue, but the API part seems to work fine. Here's what I've done:
1. checkout the develop branch
1. dropped the db
1. deployed. Ran `setup-all.sh` and `setup-users.sh` and `setup-dvs.sh`, so I can have something to work with.
1. added the localhost group `ipGroup-localhost.json`
1. Made sure Uma can't get to Pete's dataverse:
```
api$ http GET localhost:8080/api/dataverses/peteSubNormal?key=$UMA
HTTP/1.1 401 Unauthorized
Content-Length: 86
Content-Type: application/json
Date: Tue, 03 May 2016 19:22:08 GMT
Server: GlassFish Server Open Source Edition  4.1
X-Powered-By: Servlet/3.1 JSP/2.3 (GlassFish Server Open Source Edition  4.1  Java/Oracle Corporation/1.8)

{
    "message": "User @uma is not permitted to perform requested action.",
    "status": "ERROR"
}
```
1. Assigned the "Localhost connections" (`&ip/localhost`) group the "curator" privilege.
1. Ran the `GET` again:
```
api$ http GET localhost:8080/api/dataverses/peteSubNormal?key=$UMA
HTTP/1.1 200 OK
Content-Length: 418
Content-Type: application/json
Date: Tue, 03 May 2016 19:25:41 GMT
Server: GlassFish Server Open Source Edition  4.1
X-Powered-By: Servlet/3.1 JSP/2.3 (GlassFish Server Open Source Edition  4.1  Java/Oracle Corporation/1.8)

{
    "data": {
        "affiliation": "Affiliation value",
        "alias": "peteSubNormal",
        "creationDate": "2016-05-03T18:59:46Z",
        "creator": {
            "displayInfo": {
                "Title": "Pete Privileged",
                "email": "pete@malinator.com"
            },
            "identifier": "@pete"
        },
        "dataverseContacts": [
            {
                "contactEmail": "pete@mailinator.com",
                "displayOrder": 0
            }
        ],
        "description": "Where Pete stores normal data",
        "id": 3,
        "name": "Pete's public place",
        "ownerId": 2,
        "permissionRoot": true
    },
    "status": "OK"
}

```

Currently looking into the UI part.

**Comment by michbarsinai** [comment]
Log: 
1. Thanks @landreev for the super-useful comment regarding remote IP addresses in  DataverseRequest.java.
2. Remote address in JSF is obtained via `FacesContext.getCurrentInstance().getExternalContext().getRequest()`; it is not injected using `@Context`.

**Comment by michbarsinai** [comment]
Fix implemented in branch https://github.com/IQSS/dataverse/tree/1380-honor-ip-groups.

Note that dataverses available via IP groups do not appear in search. They are available using direct URL.

**Comment by kcondon** [comment]
I can now grant a role to an IP group at the dataverse level and gain direct access but dataverse card does not appear in the UI as Michael mentioned nor on My Data. Also, you do not see child cards, eg. ds, files. Last, assigning role to ds or files does not appear to allow actual access by IP groups, for example to unpublished dataset or a restricted file in a published dataset.

@michbarsinai 
I can break these other issues out into separate tickets. The main issues relating to granting perms is unpublished ds and restricted files needing permission to access cannot be accessed by IP groups.

There are a number of other issues relating to IP groups that should be addressed to fully integrate them into the app: ip groups are ready only so should you be able to assign admin role to them? What should happen: allow but only get read perms (that's what seems to happen) or not allow with error? When identified as belonging to an ip group, is there a sign in the UI to that effect?

**Comment by kcondon** [comment]
Created a ticket for the cards issue, #3105, and for the better integration of feature into the UI #3106

**Comment by michbarsinai** [comment]
Tasks:
1. Assign IP Group to an explicit group
2. Access restricted files and unpublished datasets based on IP group
1. Resync the PR

**Comment by pdurbin** [comment]
@michbarsinai I assume you don't object to #3151: Sensitive Data: IP Groups should not allow access to level 2 data or above. I'd like to at least get you thinking about this requirement.

**Comment by michbarsinai** [comment]
@pdurbin I actually do object - in the hands of a capable sysadmin, IP groups are a pretty strong and secure tool. It can be pretty hard to connect to a network socket in a locked room...
I'd do a user survey before blocking this.

**Comment by pdurbin** [comment]
@michbarsinai I created #3151 after a meeting with HUIT Security, @mcrosas and @kcondon to get answers to some questions having to do with the [Support for Sensitive Data](https://docs.google.com/document/d/1iSO6u2pRadyMbWJ-x4hXFeLEJGNuRC2ij-euWzipwzw/edit?usp=sharing) project. Let's loop you into these discussions if you object to any of the tasks being defined. It sounds like at minimum you'd like this to be configurable rather than hard coded behavior. A prerequisite for all of this, of course, is recording in the database whether data is level 2 or level 3, of course.

**Comment by michbarsinai** [comment]
@pdurbin IMHO, configurable, and defulting to "no" sounds best. IP groups are a feature for "power users" (can we still use this term?) so I assume they wouldn't mind changing an extra setting after they've configured the IP assignment table on their network.

That said, if even our power users don't want it, we can hard-code it to "no".

**Comment by djbrooke** [comment]
This came up in the 8/2 Community Call (https://docs.google.com/document/d/1FsXatbbq-F4qwD_w9dqkFXpr2Sd4MqScKevUcHLCVDY/edit).

**Comment by sbarbosadataverse** [comment]
This also came up for Numeric Data Services were anyone using Harvard Library should be able to access their dataverse via IP. Another case is "licensed" data that the PI would like other institutions that subscribe to the data, to access via Dataverse (like Princeton, etc...) This would increase use of the data itself  and bring traffic to Dataverse--both of which we want.

**Comment by michbarsinai** [comment]
While working on bug: Note that the inability to delete an IP group that has assignments is OK. It's just that the error message was misleading. Error message fixed in 2ee3fad.

**Comment by michbarsinai** [comment]
OK! off to final QA, and then, hopefully, dev.

**Comment by kcondon** [comment]
OK, basic ip group support is working. There are two remaining tickets, #3273, #3275. Closing this.
